sstables/storage.cc

```diff
@@ -41,6 +41,7 @@ namespace sstables {
 class filesystem_storage final : public sstables::storage {
     std::filesystem::path _dir;
     std::optional<std::filesystem::path> _temp_dir; // Valid while the sstable is being created, until sealed
+    mutable std::unordered_map<std::string, file> _dircache;
```

Comment:
might want to rename it to `_cached_dirs`.

 i don't get why we never evict the cached element of `_dircache`. just because we believe that the number of element should be very small?

Reply:
> might want to rename it to `_cached_dirs`.

Certainly.

> i don't get why we never evict the cached element of `_dircache`. just because we believe that the number of element should be very small?

Yes. The set of directories that this cache is supposed to hold is actually fixated at build time, using an `unordered_map` at runtime is just simpler and safer than collecting the individual directories in advance, statically (i.e., by programmer effort), and by keeping that statically collected set up-to-date as the codebase progresses.

This is how I understood the last paragraph of @bhalevy's [comment](https://github.com/scylladb/scylladb/issues/2399#issuecomment-2193823969), anyway.

Reply:
@tchaikov please confirm if this argument is acceptable to you. Thanks!

Reply:
> The set of directories that this cache is supposed to hold is actually fixated at build time.

@lersek build time? why? which are they? my concern was not using `unordered_map`, i am more interested how we *cache* the open directories. as, in general, when we design a caching system, one of the key challenges is to the cache eviction policy. but we don't evict them at all. Benny's comment ends with following words:

> All that said, let's consider also a more generic solution of caching open directory objects where we can lookup a directory by path, and if it's missing in the cache, create it, and keep it opened for later.
> Then we could consult this cache in the common paths mentioned above.

he put "for later" but not "forever".

Reply:
From Benny's earlier comments on #2399 I got the impression that this wouldn't be a generic cache, subjected to external/untrusted input, but a cache where, at least in theory, each client code location could even register itself in advance. This is what I meant by "fixated at build time". Basically all the sites that currently call `file::flush`, directly or indirectly, are known at build time, together with their effectively constant arguments (... was my understanding).

Anyway. A few days ago I thought about how I'd implement an LRU cache (and then we can cache a fixed number of directories, and evict everything else). The directory cache would be implemented by an `std::list`, owning the `file` objects (position in the list would correspond to relative age of last use; least recently used would be at the front). In addition, an unordered map would store list iterators, for lookup by directory name. Whenever a directory needs to be accessed through the cache:

- look up the list iterator through the map (by name); O(1) on average
- if found, move (splice) the referenced element to the tail of the list; O(1)
- if not found, insert a new element at the tail of the list (O(1), and add a reference to the new tail iterator to the map (by directory name) (O(1) on average). If the list is now overlong, remove the head list iterator from the map (by directory name, O(1) on avg), and drop the head of the list (O(1)).

Of course the question then becomes how large the cache should be, and whether LRU makes any sense for it at all.

Reply:
I now agree that, for any cache-by-name solution, we need eviction. After having looked more at the code, my earlier statement that the set of directories to cache was bounded at compile time now feels minimally dubious, but more plain wrong.

(I do feel that a cache-by-name interface is much more approachable to me than to track the associated `file` objects via individual C++ members and variables. The latter appears distruptive to existing interfaces (and looks way above "good first issue to solve" level, TBH.)

Reply:
resolving as obsolete per https://github.com/scylladb/scylladb/pull/19624#issuecomment-2218322451

Reply:
It was simpler before the storage abstraction was introduced. Now a directory isn't really a posix directory but instead something more abstract, but the storage interface doesn't provide a real abstraction for it. So I agree, it's quite complex, and the payoff is low.